🐛 BUG:

### Which Cloudflare product(s) does this pertain to?

C3 (npm create cloudflare)

### What version(s) of the tool(s) are you using?

2.23.0 [c3]

### What version of Node are you using?

18.20.4

### What operating system and version are you using?

Mac Sonoma 14.5

### Describe the Bug

### Observed behavior
Please describe.
create a new project with 
 category: hello world example
 lang: typescript
 git: yes
 deploy: no
immediately run 
  npm install
  npm exec wrangler dev  <-- will fail with following error

✘ [ERROR] Error reloading local server: Error: Cannot find module '@cloudflare/workers-shared/dist/asset-server-worker.mjs'


### Expected behavior
Please describe.
run dev server as usual

### Steps to reproduce
Please provide the following:
- A minimal working subset of your worker code
- A minimal working subset of your `wrangler.toml`
- Commands used to start your local dev server, including custom env and cli args
- Steps to be performed in the browser, curl commands, or a test we can run that reliably fails (at least a percent of the time)

create a new project with 
 category: hello world example
 lang: typescript
 git: yes
 deploy: no
immediately run 
  npm install
  npm exec wrangler dev  <-- will fail with following error

✘ [ERROR] Error reloading local server: Error: Cannot find module '@cloudflare/workers-shared/dist/asset-server-worker.mjs'

100% replicable

A git repo we can clone and run a test suite on, or which has a README with step-by-step instructions, is even better. In this case, please use the field below to provide a link to the minimal repro.


### Please provide a link to a minimal reproduction

_No response_

### Please provide any relevant error logs

_No response_

--- a/.changeset/quiet-lizards-play.md
+++ b/.changeset/quiet-lizards-play.md
@@ -1,0 +1,5 @@
+---
+"wrangler": patch
+---
+
+fix: prevent crash when running wrangler dev due to missing dependency

--- a/packages/wrangler/package.json
+++ b/packages/wrangler/package.json
@@ -70,6 +70,7 @@
 	},
 	"dependencies": {
 		"@cloudflare/kv-asset-handler": "workspace:*",
+		"@cloudflare/workers-shared": "workspace:*",
 		"@esbuild-plugins/node-globals-polyfill": "^0.2.3",
 		"@esbuild-plugins/node-modules-polyfill": "^0.2.2",
 		"blake3-wasm": "^2.1.5",
@@ -92,7 +93,6 @@
 		"@cloudflare/eslint-config-worker": "workspace:*",
 		"@cloudflare/pages-shared": "workspace:^",
 		"@cloudflare/types": "^6.18.4",
-		"@cloudflare/workers-shared": "workspace:*",
 		"@cloudflare/workers-tsconfig": "workspace:*",
 		"@cloudflare/workers-types": "^4.20240725.0",
 		"@cspotcode/source-map-support": "0.8.1",

--- a/packages/wrangler/src/dev/miniflare.ts
+++ b/packages/wrangler/src/dev/miniflare.ts
@@ -842,43 +842,6 @@ export async function buildMiniflareOptions(
 		}
 	}
 
-	const assetServerModulePath = require.resolve(
-		"@cloudflare/workers-shared/dist/asset-server-worker.mjs"
-	);
-	const assetServerConfigPath = require.resolve(
-		"@cloudflare/workers-shared/asset-server-worker/wrangler.toml"
-	);
-	let assetServerConfig: Config | undefined;
-
-	try {
-		assetServerConfig = readConfig(assetServerConfigPath, {});
-	} catch (err) {
-		throw new UserError(
-			"Failed to read the Asset Server Worker configuration file.\n" + `${err}`
-		);
-	}
-
-	const assetServerWorker: WorkerOptions | undefined = config.experimentalAssets
-		? {
-				name: assetServerConfig?.name,
-				compatibilityDate: assetServerConfig?.compatibility_date,
-				compatibilityFlags: assetServerConfig?.compatibility_flags,
-				modulesRoot: dirname(assetServerModulePath),
-				modules: [
-					{
-						type: "ESModule",
-						path: assetServerModulePath,
-					},
-				],
-				unsafeDirectSockets: [
-					{
-						host: "127.0.0.1",
-						port: 0,
-					},
-				],
-			}
-		: undefined;
-
 	const upstream =
 		typeof config.localUpstream === "string"
 			? `${config.upstreamProtocol}://${config.localUpstream}`
@@ -921,13 +884,55 @@ export async function buildMiniflareOptions(
 					proxy: true,
 				})),
 			},
-			...(config.experimentalAssets
-				? [assetServerWorker as WorkerOptions]
-				: []),
+			...getAssetServerWorker(config),
 			...externalWorkers,
 		],
 	};
 	return { options, internalObjects, entrypointNames };
+}
+
+function getAssetServerWorker(
+	config: Omit<ConfigBundle, "rules">
+): WorkerOptions[] {
+	if (!config.experimentalAssets) {
+		return [];
+	}
+	const assetServerModulePath = require.resolve(
+		"@cloudflare/workers-shared/dist/asset-server-worker.mjs"
+	);
+	const assetServerConfigPath = require.resolve(
+		"@cloudflare/workers-shared/asset-server-worker/wrangler.toml"
+	);
+	let assetServerConfig: Config | undefined;
+
+	try {
+		assetServerConfig = readConfig(assetServerConfigPath, {});
+	} catch (err) {
+		throw new UserError(
+			"Failed to read the Asset Server Worker configuration file.\n" + `${err}`
+		);
+	}
+
+	return [
+		{
+			name: assetServerConfig?.name,
+			compatibilityDate: assetServerConfig?.compatibility_date,
+			compatibilityFlags: assetServerConfig?.compatibility_flags,
+			modulesRoot: dirname(assetServerModulePath),
+			modules: [
+				{
+					type: "ESModule",
+					path: assetServerModulePath,
+				},
+			],
+			unsafeDirectSockets: [
+				{
+					host: "127.0.0.1",
+					port: 0,
+				},
+			],
+		},
+	];
 }
 
 export interface ReloadedEventOptions {

